syntax = "proto3";

package redpepper;

enum MessageType {
    UNSPECIFIED = 0;
    CLIENTHELLO = 10;
    SERVERHELLO = 11;
    BYE = 12;
    PING = 13;
    PONG = 14;

    COMMAND = 20;
    COMMANDCANCEL = 21;
    COMMANDSTATUS = 22;

    DATAREQUEST = 30;
    DATARESPONSE = 31;
}

message Message {
    MessageType type = 1;
    oneof message {
        ClientHello client_hello = 10;
        ServerHello server_hello = 11;
        Bye bye = 13;
        Ping ping = 14;
        Pong pong = 15;

        Command command = 20;
        CommandCancel cancel = 21;
        CommandStatus status = 22;

        DataRequest data_request = 30;
        DataResponse data_response = 31;
    }
}

message Bye {
    string reason = 1;
}

message ClientHello {
    string clientID = 1;
    string auth = 2;
}

message ServerHello {
    uint32 version = 1;
}

message Ping {
    int32 data = 1;
}

message Pong {
    int32 data = 1;
}

message Command {
    int32 commandID = 1;
    string type = 2;
    string data = 3;
}

message CommandCancel {
    int32 commandID = 1;
}

message CommandStatus {
    int32 commandID = 1;
    enum Status {
        PENDING = 0;
        SUCCESS = 1;
        FAILED = 2;
        CANCELLED = 3;
    }
    Status status = 2;
    Progress progress = 3;
    string data = 4;
}

message Progress {
    uint32 current = 1;
    uint32 total = 2;
}

message DataRequest {
    int32 requestID = 1;
    string type = 2;
    string data = 3;
}

message DataResponse {
    int32 requestID = 1;
    bool ok = 2;
    string data = 3;
}

